images/upload-gitlab-failure-logs/upload_gitlab_failure_logs.py

```diff
@@ -4,36 +4,64 @@
 import re
 from pathlib import Path
 
-import gitlab
 import yaml
-from opensearch_dsl import Date, Document, connections
 
+GITLAB_TOKEN = os.environ.get("GITLAB_TOKEN")
+OPENSEARCH_ENDPOINT = os.environ.get("OPENSEARCH_ENDPOINT")
+OPENSEARCH_USERNAME = os.environ.get("OPENSEARCH_USERNAME")
+OPENSEARCH_PASSWORD = os.environ.get("OPENSEARCH_PASSWORD")
 
-class JobPayload(Document):
-    timestamp = Date()
 
-    class Index:
-        name = "gitlab-job-failures-*"
+def load_taxonomy():
+    with open(Path(__file__).parent / "taxonomy.yaml") as f:
+        data = yaml.safe_load(f)["taxonomy"]
+    return data
+
 
-    def save(self, **kwargs):
-        # assign now if no timestamp given
-        if not self.timestamp:
-            self.timestamp = datetime.now()
+taxonomy = load_taxonomy()
 
-        # override the index to go to the proper timeslot
-        kwargs["index"] = self.timestamp.strftime("gitlab-job-failures-%Y%m%d")
-        return super().save(**kwargs)
 
+def classify(job_trace: str):
+    job_error_class = None
+    matching_patterns = set()
+    for error_class, lookups in taxonomy["error_classes"].items():
+        if lookups:
+            for grep_expr in lookups.get("grep_for", []):
+                if re.compile(grep_expr).search(job_trace):
+                    matching_patterns.add(error_class)
+
+    # If the job logs matched any regexes, assign it the taxonomy
+    # with the highest priority in the "deconflict order".
+    # Otherwise, assign it a taxonomy of "other".
+    if len(matching_patterns):
+        for error_class in taxonomy["deconflict_order"]:
+            if error_class in matching_patterns:
+                job_error_class = error_class
+                break
+    else:
+        job_error_class = "other"
 
-GITLAB_TOKEN = os.environ["GITLAB_TOKEN"]
-OPENSEARCH_ENDPOINT = os.environ["OPENSEARCH_ENDPOINT"]
-OPENSEARCH_USERNAME = os.environ["OPENSEARCH_USERNAME"]
-OPENSEARCH_PASSWORD = os.environ["OPENSEARCH_PASSWORD"]
+    return job_error_class
```

Comment:
Following https://github.com/spack/spack-infrastructure/pull/440#issuecomment-1490634469, why can't we:
```suggestion
    return matching_patterns, job_error_class
```
?